## Data Frame Creation and Manipulation
1.	Create a data frame with details of 10 students and columns as Roll Number, Name, Gender, Marks1, Marks2, Marks3.
2.	Create a new column with total marks
3.	Find the lowest marks in Marks1
4.	Find the Highest marks in Marks2
5.	Find the average marks in Marks3
6.	Find student name with highest average
7.	Find how many students failed in Marks2 (<40)


First, we download the python libraries we will be using in this notebook. The libraries we have used are:
*   NumPy - for data manipulation
*   Pandas - for data frame creation and manipulation
*   MatPlotLib - for data visualization
*   Seaborn - for data visualization


In [4]:
# importing required libraries
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn

Next, we start solving the questions in the lab.


#### **Q1.** Create a data frame with details of 10 students and columns as Roll Number, Name, Gender, Marks1, Marks2, Marks3.

To create a dataframe, we have first created a dictionary student_data with the following keys:     
*   Roll_Number: a list of 10 integers which represent the roll nos. of 10 students in the class.
*   Name: a list of 10 strings which represent the names of the 10 students in the class.
*   Gender: a list of 10 characters which represent the genders of the 10 students in the class. Here, F stands for Female and M stands for Male.
*   Marks1: a list of 10 integers which represent the scores the 10 students have scored in Subject 1. Here, the minimum marks a student can score is 0, and the maximum marks a student can score is 100.
*   Marks2: a list of 10 integers which represent the scores the 10 students have scored in Subject 2. Here, the minimum marks a student can score is 0, and the maximum marks a student can score is 100.
*   Marks3: a list of 10 integers which represent the scores the 10 students have scored in Subject 3. Here, the minimum marks a student can score is 0, and the maximum marks a student can score is 100.


After creating the dictionary, we used the DataFrame() method from the Pandas library to convert the dictionary to a dataframe. Here, the keys become the column names and the value lists occupy the rows of the dataframe.

In [6]:
# 1. Creating Dataframe

# creating the dictionary
student_data = {
    
    # key Roll_Number to store the roll numbers of the students
    'Roll_Number' : [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],

    # key Name to store the names of the students
    'Name' : ['Monica', 'Rachel', 'Phoebe', 'Ross', 'Joey', 'Chandler', 'Jack', 'Judy', 'Emily', 'Richard'],

    # key Gender to store the genders of the students
    'Gender' : ['F', 'F', 'F', 'M', 'M', 'M', 'M', 'F', 'F', 'M'],

    # key Marks1 to store the marks earned by the students in Subject1; Min Marks = 0, Max Marks = 100.
    'Marks1' : [56, 90, 78, 66, 13, 67, 83, 22, 100, 94],

    # key Marks2 to store the marks earned by the students in Subject2; Min Marks = 0, Max Marks = 100.
    'Marks2' : [44, 89, 71, 23, 100, 87, 92, 65, 55, 43],

    # key Marks3 to store the marks earned by the students in Subject3; Min Marks = 0, Max Marks = 100.
    'Marks3' : [75, 18, 90, 56, 83, 21, 97, 9, 100, 44]
}

# converting the dictionary to a dataframe
df = pd.DataFrame(student_data)

After creating the dataframe, we view it to check whether the dataframe has been created successfully.

In [12]:
# viewing the dataframe
df

Roll_Number      Name Gender  Marks1  Marks2  Marks3
0            1    Monica      F      56      44      75
1            2    Rachel      F      90      89      18
2            3    Phoebe      F      78      71      90
3            4      Ross      M      66      23      56
4            5      Joey      M      13     100      83
5            6  Chandler      M      67      87      21
6            7      Jack      M      83      92      97
7            8      Judy      F      22      65       9
8            9     Emily      F     100      55     100
9           10   Richard      M      94      43      44

#### **Q2.** Create a new column with total marks.

To solve Question 2, we create a new column in the dataframe to store the total marks earned by a student. To do this, we add the three columns - Marks1, Marks2, and Marks3 and store the result in a column called Total_Marks.

In [31]:
# 2. Creating Columm Total Marks

# creating a new column Total_Marks to store the sum of data stored in columns Marks1, Marks2, and Marks3
df['Total_Marks'] = df.Marks1 + df.Marks2 + df.Marks3

# viewing the dataframe
df

Roll_Number      Name Gender  Marks1  Marks2  Marks3  Total_Marks
0            1    Monica      F      56      44      75          175
1            2    Rachel      F      90      89      18          197
2            3    Phoebe      F      78      71      90          239
3            4      Ross      M      66      23      56          145
4            5      Joey      M      13     100      83          196
5            6  Chandler      M      67      87      21          175
6            7      Jack      M      83      92      97          272
7            8      Judy      F      22      65       9           96
8            9     Emily      F     100      55     100          255
9           10   Richard      M      94      43      44          181

#### **Q3.** Find the lowest marks in Marks1.

To solve Question 3, we use the min() function from the Pandas library on the column Marks1 of the dataframe.



In [33]:
# 3. Finding lowest marks in Marks1
df.Marks1.min()

13

As we can see, the lowest marks in Marks1 is 13.

#### **Q4.** Find the Highest marks in Marks2.

To solve Question 4, we use the max() function from the Pandas library on the column Marks2 of the dataframe.

In [35]:
# 4. Finding highest marks in Marks2
df.Marks2.max()

100

As we can see, the highest marks in Marks2 is 100.

#### **Q5.** Find the average marks in Marks3.

To solve Question 5, we use the mean() function from the Pandas library on the column Marks3 of the dataframe.

In [37]:
# 5. Finding average marks in Marks3
df.Marks3.mean()

59.3

As we can see, the average marks in Marks3 is 59.3.

#### **Q6.** Find student name with highest average.

To solve Question 6, we filter the dataframe with the condition that the Total_Marks observation matches the maximum value of the Total_Marks. In this filtered dataframe, we check the observation in the Name column. Note: the student with the highest marks will have the highest average as average = total/3.

In [39]:
# 6. Finding student with highest average
df[df.Total_Marks==df.Total_Marks.max()].Name

6    Jack
Name: Name, dtype: object

As we can see, the student with the highest average (the student with the highest total will have the highest average as average=total/3) is Jack.

#### **Q7.** Find how many students failed in Marks2 (<40).

To solve Question 7, we filter the dataframe with the condition that the Marks2 observation is less than 40. Then we apply the len() function to the filtered dataframe to find the length/ number of rows/ number of students who failed.

In [41]:
# 7. Finding number of students who failed in Marks2
len(df[df.Marks2<40])

1

As we can see, the number of students who failed in Marks2 (scored less than 40) is 1.